File: src/advent/2020/day5.clj
(ns advent.2020.day5
  (:require [advent.core :as a]
            [clojure.set :as sets]))

(def reader (a/read-input 2020 5))

(def input (reader))

(defn bsp-> [code init]
  (loop [[min max] init
         [c & tail] code]
    (if (or (= min max)
            (nil? c))
      min
      (let [distance (inc (- max min))
            half (->> (/ distance 2)
                      (int)
                      (+ min))
            range (case c
                    \F [min (dec half)]
                    \L [min (dec half)]
                    \B [half max]
                    \R [half max])]
        (recur range tail)))))

(defn seat [code]
  (let [row (subs code 0 7)
        column (subs code 7)]
    [(bsp-> row [0 127]) (bsp-> column [0 7])]))

(defn seat-id [[r c]]
  (->> (* r 8)
       (+ c)))

(defn code->seat-id [code]
  (seat-id (seat code)))

(defn part1 []
  (->> input
       (map code->seat-id)
       (apply max)))


(defn near-seats [ids id]
  (and (contains? ids (inc id))
       (contains? ids (dec id))))
(defn part2 []
  (let [seats (->> input
                   (map seat)
                   (set))
        ids (set (map seat-id seats))
        all (for [row (range 0 128)
                  col (range 0 8)]
              [row col])]
    (->> all
         (filter #(not (contains? seats %)))
         (map seat-id)
         (filter #(and (contains? ids (inc %))
                       (contains? ids (dec %)))))))

File: zig/src/2020/day5.zig
const std = @import("std");
const advent = @import("advent.zig");

pub fn part1(allocator: std.mem.Allocator, variant: []const u8) !i32 {
    var input = try advent.DayOfAdvent.init(allocator, "2020", "day5", variant);
    defer input.deinit();

    var buffer: [512]u8 = undefined;
    var reader = input.reader();

    var higher: i32 = 0;
    while (try reader.readUntilDelimiterOrEof(buffer[0..], '\n')) |line| {
        const seat = try getSeat(line);
        const id = seat.id();
        if (id > higher) {
            higher = id;
        }
    }
    return higher;
}

pub fn part2(allocator: std.mem.Allocator, variant: []const u8) !i32 {
    var input = try advent.DayOfAdvent.init(allocator, "2020", "day5", variant);
    defer input.deinit();

    var buffer: [512]u8 = undefined;
    var reader = input.reader();

    var seats = std.AutoHashMap(i32, void).init(allocator);
    defer seats.deinit();

    while (try reader.readUntilDelimiterOrEof(buffer[0..], '\n')) |line| {
        const seat = try getSeat(line);
        const id = seat.id();
        try seats.put(id, {});
    }
    for (1..126) |r| {
        for (0..7) |c| {
            const id: i32 = @intCast(r * 8 + c);
            if (!seats.contains(id) and seats.contains(id - 1) and seats.contains(id + 1)) {
                return id;
            }
        }
    }
    return -1;
}
const Seat = struct {
    row: u8,
    column: u8,
    fn id(self: Seat) i32 {
        return (@as(i32, self.row) * 8) + self.column;
    }
};
fn getSeat(line: []const u8) !Seat {
    if (line.len != 10) {
        return error.InvalidSyntax;
    }
    return .{.row = binarySpace(line[0..7], 'B', 128), .column = binarySpace(line[7..10], 'R', 8)};
}
fn binarySpace(string: []const u8, upper: u8, space: u8) u8 {
    var min: u8 = 0;
    var max: u8 = space;
    for (string[0..]) |c| {
        min, max = binarySpaceStep(c == upper, min, max);
    }
    return min;
}
test "binarySpace" {
    const row = "FBFBBFF";
    var res = binarySpace(row, 'B', 128);
    try std.testing.expectEqual(44, res);
    const col = "RLR";
    res = binarySpace(col, 'R', 8);
    try std.testing.expectEqual(5, res);
}
fn binarySpaceStep(upperBound: bool, min: u8, max: u8) [2]u8 {
    const half = min + ((max - min) / 2);
    if (upperBound) {
        return .{half, max};
    } else {
        return .{min, half};
    }
}

test "Test example" {
    try std.testing.expectEqual(820, part1(std.testing.allocator, "example.txt"));
}

test "part 1" {
    try std.testing.expectEqual(928, part1(std.testing.allocator, "input.txt"));
}
test "part 2" {
    try std.testing.expectEqual(610, part2(std.testing.allocator, "input.txt"));
}
